storagenode/piecestore/endpoint.go

```diff
@@ -551,6 +553,50 @@ func (endpoint *Endpoint) SaveOrder(ctx context.Context, limit *pb.OrderLimit, o
 	}
 }
 
+// Retain keeps only piece ids specified in the request
+func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainRequest) (*pb.RetainResponse, error) {
+	peer, err := identity.PeerIdentityFromContext(ctx)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	filter, err := bloomfilter.NewFromBytes(retainReq.GetFilter())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	limit := 20
+	offset := 0
+	piecesRemaining := true
+
+	for piecesRemaining {
+		pieceIDs, err := endpoint.pieceinfo.GetPiecesID(ctx, peer.ID, retainReq.GetCreationDate(), limit, offset)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+		for _, pieceID := range pieceIDs {
+			if !filter.Contains(pieceID) {
+				err = errs.Combine(err, endpoint.store.Delete(ctx, peer.ID, pieceID))
+				err = errs.Combine(err, endpoint.pieceinfo.Delete(ctx, peer.ID, pieceID))
+			}
+		}
+		piecesRemaining = (len(pieceIDs) == limit)
+		offset += limit
+		runtime.Gosched()
+	}
+	return &pb.RetainResponse{}, Error.Wrap(err)
```

Comment:
If we return an error here, I think it will be returned as "internal error" back to the satellite. I am not sure if the satellite is interested if the storage node failed to collect some of the garbage... Perhaps, we should focus on logging for the SNO.

Reply:
does this apply only to this line or throughout this function?